File: backend/recognizer/fen.py
"""
棋盘矩阵 → FEN 字符串。

严格对齐前端约定 (src/engine/fen.ts, src/types/index.ts):
  - board[rank][file]，rank 0 = 黑方底线 (图像最上一行)，rank 9 = 红方底线 (图像最下一行)
  - file 0..8 从红方左侧数
  - WXF 风格字符: K(帅/将) A(仕/士) E->B(相/象) H->N(马) R(车) C(炮) P(兵/卒)
  - 大写 = 红方, 小写 = 黑方
  - FEN board 部分: rank 0 → rank 9，每行 file 0 → file 8

棋子内部表示: (type, color)
  type ∈ {K, A, E, H, R, C, P}
  color ∈ {'red', 'black'}
  空格用 None
"""

from __future__ import annotations

from typing import List, Optional, Tuple

FILES = 9
RANKS = 10

# 内部 type → FEN 字符 (大写, WXF 风格)
_TYPE_TO_FEN = {
    "K": "K",
    "A": "A",
    "E": "B",  # 相/象
    "H": "N",  # 马
    "R": "R",
    "C": "C",
    "P": "P",
}

# 棋子: (type, color) 或 None
Piece = Optional[Tuple[str, str]]
Board = List[List[Piece]]


def empty_board() -> Board:
    return [[None for _ in range(FILES)] for _ in range(RANKS)]


def piece_char(piece: Tuple[str, str]) -> str:
    ptype, color = piece
    ch = _TYPE_TO_FEN[ptype]
    return ch if color == "red" else ch.lower()


def board_to_fen(
    board: Board,
    side: str = "red",
    halfmove: int = 0,
    fullmove: int = 1,
) -> str:
    """把 9x10 棋盘矩阵转成完整 FEN 串。"""
    if len(board) != RANKS or any(len(row) != FILES for row in board):
        raise ValueError(f"board 形状必须是 {RANKS}x{FILES}")

    rank_strings: List[str] = []
    for r in range(RANKS):
        row = ""
        empty = 0
        for f in range(FILES):
            sq = board[r][f]
            if sq is None:
                empty += 1
            else:
                if empty > 0:
                    row += str(empty)
                    empty = 0
                row += piece_char(sq)
        if empty > 0:
            row += str(empty)
        rank_strings.append(row)

    side_char = "w" if side == "red" else "b"
    return f"{'/'.join(rank_strings)} {side_char} - - {halfmove} {fullmove}"

File: src/engine/fen.ts
/**
 * FEN (Forsyth-Edwards Notation) for Chinese Chess
 * 标准 WXF/UCCI 风格 FEN
 *
 * 格式: <board> <side> <en-passant> <halfmove> <fullmove>
 * 例: "rnbakabnr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RNBAKABNR w - - 0 1"
 *
 * board: rank 9 (黑底线) → rank 0 (红底线)，每行从 file 0 开始
 * 大写 = 红方，小写 = 黑方
 */

import type { Board, Color, Piece, Square } from '@/types';
import { FILES, RANKS, emptyBoard, pieceChar } from '@/types';

const PIECE_CHARS: Record<string, Piece> = {
  K: { type: 'K', color: 'red' },
  A: { type: 'A', color: 'red' },
  E: { type: 'E', color: 'red' },
  B: { type: 'E', color: 'red' }, // 别名: 某些 FEN 用 B 表示相 (Bishop 风格)
  N: { type: 'H', color: 'red' }, // 别名: WXF 风格 FEN 用 N 表示马 (knight 风格)
  H: { type: 'H', color: 'red' },
  R: { type: 'R', color: 'red' },
  C: { type: 'C', color: 'red' },
  P: { type: 'P', color: 'red' },
  k: { type: 'K', color: 'black' },
  a: { type: 'A', color: 'black' },
  e: { type: 'E', color: 'black' },
  b: { type: 'E', color: 'black' }, // 别名
  n: { type: 'H', color: 'black' }, // 别名
  h: { type: 'H', color: 'black' },
  r: { type: 'R', color: 'black' },
  c: { type: 'C', color: 'black' },
  p: { type: 'P', color: 'black' },
};

export interface FenState {
  board: Board;
  side: Color;
  halfmove: number; // 自上次吃子/动兵以来的半回合数
  fullmove: number; // 总回合数
}

export function parseFen(fen: string): FenState {
  const [boardPart, sidePart = 'w', , , halfmovePart = '0', fullmovePart = '1'] =
    fen.trim().split(/\s+/);

  const board = emptyBoard();
  const ranks = boardPart.split('/');
  if (ranks.length !== RANKS) {
    throw new Error(`Invalid FEN: expected ${RANKS} ranks, got ${ranks.length}`);
  }

  for (let r = 0; r < RANKS; r++) {
    // FEN rank 0 是黑底线 → 我们内部 board 索引 0 也是黑底线
    const row = ranks[r];
    let file = 0;
    for (const ch of row) {
      if (/[1-9]/.test(ch)) {
        file += parseInt(ch, 10);
      } else if (PIECE_CHARS[ch]) {
        board[r][file] = PIECE_CHARS[ch];
        file++;
      } else {
        throw new Error(`Invalid FEN char at rank ${r}, file ${file}: ${ch}`);
      }
    }
    if (file !== FILES) {
      throw new Error(`Invalid FEN: rank ${r} has ${file} files, expected ${FILES}`);
    }
  }

  return {
    board,
    side: sidePart === 'b' ? 'black' : 'red',
    halfmove: parseInt(halfmovePart, 10) || 0,
    fullmove: parseInt(fullmovePart, 10) || 1,
  };
}

export function toFen(state: FenState): string {
  const rankStrings: string[] = [];
  for (let r = 0; r < RANKS; r++) {
    let row = '';
    let empty = 0;
    for (let f = 0; f < FILES; f++) {
      const sq = state.board[r][f];
      if (sq === null) {
        empty++;
      } else {
        if (empty > 0) {
          row += String(empty);
          empty = 0;
        }
        row += pieceChar(sq);
      }
    }
    if (empty > 0) row += String(empty);
    rankStrings.push(row);
  }
  return `${rankStrings.join('/')} ${state.side === 'red' ? 'w' : 'b'} - - ${state.halfmove} ${state.fullmove}`;
}

export function cloneBoard(board: Board): Board {
  return board.map((rank) => rank.map((sq) => (sq ? { ...sq } : null)));
}

export function squareAt(
  board: Board,
  file: number,
  rank: number,
): Square {
  return board[rank][file];
}